### PR TITLE
Adding image downloading

### PR DESCRIPTION
**Pull request by danielfaulknor**
Replaces #18

Adding image downloads as in #8

Simply adds in a function to the filter to download the media in the tweet. Assumes you're embedding images

It's working well for me.

### CODE REVIEW

File: inc/import.php [hunk 1]
```diff
@@ -199,6 +199,41 @@ function ozh_ta_linkify_tweet( $tweet ) {
             $display_url  = $media->display_url;
             $expanded_url = $media->expanded_url;
             $tco_url      = $media->url;
+
+	    if( $ozh_ta['download_images'] == 'yes' ) {
+                $url = $media_url;
+    		$tmp = download_url( $url );
+    		
+		if( is_wp_error( $tmp ) ){
+           	    // download failed, handle error
```
Comment (kw217) [inline inc/import.php]: This block does nothing. Instead, it should have an else block enclosing the rest of the code (up to line 235). If we haven't downloaded it there's no point attempting to store into `$file_array` then taking it out again.

Is there a way of reporting failures to the user, maybe by incrementing a counter visible in the theme or on the settings page?

File: inc/import.php [hunk 2]
```diff
@@ -199,6 +199,41 @@ function ozh_ta_linkify_tweet( $tweet ) {
...
+    		}
+
+	        $file_array = array();
+
+    	        // Set variables for storage
+    	        // fix file filename for query strings
+    	        preg_match('/[^\?]+\.(jpg|jpe|jpeg|gif|png)/i', $url, $matches);
+                $file_array['name'] = basename($matches[0]);
```
Comment (kw217) [inline inc/import.php]: Indentation not quite right.

File: inc/import.php [hunk 3]
```diff
@@ -199,6 +199,41 @@ function ozh_ta_linkify_tweet( $tweet ) {
             $display_url  = $media->display_url;
             $expanded_url = $media->expanded_url;
             $tco_url      = $media->url;
+
+	    if( $ozh_ta['download_images'] == 'yes' ) {
+                $url = $media_url;
```
Comment (kw217) [inline inc/import.php]: Indentation.

**Comment by kw217** [comment]
I've tested this and it works. Installed, manually archived a tweet, checked the image was hosted on my blog, turned off downloading in config, manually archived a tweet, checked the image was hosted on Twitter. Thanks!

I've not tested the case where the image download fails, though, and the code doesn't look quite right there. It probably works, but it would be good to clean it up and test somehow.

**Comment by kw217** [comment]
(@ozh asked me to test this for you :-) )

**Comment by ozh** [comment]
@kw217 thanks for reporting! Indeed, there needs to be something done in case the image download fails or times out

**Comment by kw217** [comment]
I was thinking about this last night. How about adding a postmeta entry for the post on failure (say `ozh_ta_err_dl`: `1`), recording the fact that there was a failed download? Then the config page can display a count of posts with failed downloads. For extra points, you could add a button to retry them - but even just tracking them would be useful enough.

**Comment by danielfaulknor** [comment]
Thanks, I'll have a crack at it if I get time over the holiday season

**Comment by visual23** [comment]
@kw217 

Thanks for writing this! I added it to my import.php but when I run it, I get this error.
Fatal error: Uncaught Error: Call to undefined function download_url() 

I don't see a 'download_url' function anywhere in the code.

Can you help?

Thanks!

**Comment by visual23** [comment]
Never mind! I needed to require these at the top of import.php.

require_once( ABSPATH . 'wp-admin/includes/media.php');
require_once( ABSPATH . 'wp-admin/includes/file.php');
require_once( ABSPATH . 'wp-admin/includes/image.php’);

**Pull request closed without merging**